### PR TITLE
Allows signing with self-signed certificate

### PR DESCRIPTION
**Pull request by priteshbandi**
Implements https://github.com/notaryproject/notaryproject/pull/193

Signed-off-by: Pritesh Bandi <pritesb@amazon.com>

### CODE REVIEW

File: testhelper/certificatetest.go [hunk 1]
```diff
@@ -139,10 +159,8 @@ func getCertTemplate(isRoot bool, cn string) *x509.Certificate {
 		template.MaxPathLen = 1
 		template.IsCA = true
 	} else {
-		template.SerialNumber = big.NewInt(2)
+		template.SerialNumber = big.NewInt(int64(mrand.Intn(200)))
 		template.NotAfter = time.Now().AddDate(0, 0, 1)
-		template.KeyUsage = x509.KeyUsageDigitalSignature
-		template.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning}
```
Comment (patrickzheng200) [inline testhelper/certificatetest.go:145]: Why are we removing these two lines?

File: testhelper/certificatetest.go [hunk 2]
```diff
@@ -62,6 +64,12 @@ func GetUnsupportedCertificate() RSACertTuple {
 	return unsupported
 }
 
+// GetRSASelfSignedSigningCertificate returns a self-signed certificate created which can be used for signing.
+// which is not supported by notary.
```
[the remark this answers is not in the document]
Comment (priteshbandi) [reply testhelper/certificatetest.go:68]: Oops that's a copy paste error will remove `which is not supported by notary.`.

File: testhelper/certificatetest.go (hunk 1 above)
Comment (priteshbandi) [reply testhelper/certificatetest.go:145]: KU and EKU are already added in line 151 and 152

File: x509/cert_validations.go [hunk 3]
```diff
@@ -25,8 +25,17 @@ func ValidateTimeStampingCertChain(certChain []*x509.Certificate, signingTime ti
 }
 
 func validateCertChain(certChain []*x509.Certificate, expectedLeafEku x509.ExtKeyUsage, signingTime time.Time) error {
-	if len(certChain) < 2 {
-		return errors.New("certificate chain must contain at least two certificates: a root and a leaf certificate")
+	if len(certChain) < 1 {
+		return errors.New("certificate chain must contain at least one certificates")
```
Comment (jondonas) [inline x509/cert_validations.go:29]: Nit: `one certificate`

File: x509/cert_validations.go [hunk 4]
```diff
@@ -25,8 +25,17 @@ func ValidateTimeStampingCertChain(certChain []*x509.Certificate, signingTime ti
...
+	}
+
+	// For self-signed signing certificate (not a CA)
+	if len(certChain) == 1 {
+		cert := certChain[0]
+		if err := cert.CheckSignature(cert.SignatureAlgorithm, cert.RawTBSCertificate, cert.Signature); err != nil {
```
Comment (jondonas) [inline x509/cert_validations.go:35]: Can we use `isSelfSigned` method below?

File: x509/cert_validations.go [hunk 5]
```diff
@@ -25,8 +25,17 @@ func ValidateTimeStampingCertChain(certChain []*x509.Certificate, signingTime ti
...
+			return err
+		}
+		return validateLeafCertificate(cert, expectedLeafEku)
```
Comment (jondonas) [inline x509/cert_validations.go:38]: Do we care about any other validations such as cert expiry?

File: x509/cert_validations.go (hunk 4 above)
Comment (priteshbandi) [reply x509/cert_validations.go:35]: We cannot use `isSelfSigned` as it internally calls `subject.CheckSignatureFrom(issuer)` that performs BasicConstrant and KeyUsageCertSign validations which are not valid for self signed leaf/signing certificate.

File: x509/cert_validations.go (hunk 5 above)
Comment (priteshbandi) [reply x509/cert_validations.go:38]: Good catch, will add that.

**Review by jondonas — APPROVED** [review]

File: x509/cert_validations_test.go [hunk 6]
```diff
@@ -158,6 +160,39 @@ var codeSigningLeafInvalidPathLenPem = "-----BEGIN CERTIFICATE-----\n" +
 	"L0Z9ZU3NwwFAh1wpSsoQR4pPN+ZkVj5Irr/KWFk=\n" +
 	"-----END CERTIFICATE-----"
 
+var openSSLMinimumPem = "-----BEGIN CERTIFICATE-----\n" +
```
Comment (shizhMSFT) [inline x509/cert_validations_test.go:163]: nit: we can use backquotes (see [string literals](https://go.dev/ref/spec#String_literals)) or embedding files (see [embed](https://pkg.go.dev/embed)).

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**